### PR TITLE
State Summary (or: `show()` methods for everyone!)

### LINKED ISSUE #175: Testing convergence
Could we have something similar to `converged` function from Optim.jl? See here: https://github.com/JuliaNLSolvers/Optim.jl/blob/master/src/api.jl#L79 . Optimization.jl would like to have such information.

### PR DESCRIPTION
**Pull request by kellertuer**
I had an idea how to introduce more informative information from a solver run – when one looks at the state (i.e. uses `return_state=true`).
This mainly involves writing quite a few `show()` methods.

Here is three examples

## 1 Just looking at a state

```Julia
using Manifolds, Manopt
gst = GradientDescentState(Euclidean(2), [0.0,0.0]; stopping_criterion=StopAfterIteration(20), stepsize=ConstantStepsize(1.0))
```
prints
```JuliaREPL
# Solver state for gradient_descent

## Parameters
* retraction method: ExponentialRetraction()

## Stepsize
ConstantStepsize{Float64}(1.0)

## Stopping Criterion
Max Iteration 20:       not reached
This indicates convergence: No
```

## 2 A solver run that was succesfull

```Julia
using Manifolds, Manopt
p = [4.0, 2.0]
f(M, q) = distance(M, q, p) .^ 2
grad_f(M, q) = -2 * log(M, q, p)
s = gradient_descent(Euclidean(2), f, grad_f, zero(p); return_state=true)
```

prints

```JuliaREPL
# Solver state for gradient_descent
After 70 iterations

## Parameters
* retraction method: ExponentialRetraction()

## Stepsize
ArmijoLinesearch{ExponentialRetraction, typeof(Manopt.armijo_initial_guess)}(1.0, ExponentialRetraction(), 0.95, 0.1, 0.8573749999999999, 0.0, Manopt.armijo_initial_guess)

## Stopping Criterion
Stop When _one_ of the following are fulfilled:
    Max Iteration 200:  not reached
    |Δf| < 1.0e-9: reached
Overall: reached
This indicates convergence: Yes
```

## 3 A solver run that was not sucessful

```julia
s2 = gradient_descent(Euclidean(2), f, grad_f, zero(p); return_state=true, stopping_criterion=StopAfterIteration(2))
```

which prints

```
# Solver state for gradient_descent
After 2 iterations

## Parameters
* retraction method: ExponentialRetraction()

## Stepsize
ArmijoLinesearch{ExponentialRetraction, typeof(Manopt.armijo_initial_guess)}(1.0, ExponentialRetraction(), 0.95, 0.1, 0.8573749999999999, 0.0, Manopt.armijo_initial_guess)

## Stopping Criterion
Max Iteration 2:        reached
This indicates convergence: No
````

# ToDo
* [x] show-methods for stopping criteria
* [x] store and obtain the iteration a stopping criterion indicated to stop (resolves #175)
* [x] introduce a `get_count` function that could also later be used on Problems maybe? Not yet sure how to access a Problem after a run. 
* [x] `show()`-Methods for
  * [x] DebugActions 
  * [x] RecordActions
  * [x] linesearches 
  * [x] States
    * [x] Gradient Descent
    * [x] Conjugate Gradient descent
      * [x] show methods for the CG coefficients
    * [x] Alternating Gradient Descent
    * [x] Augmented Lagrangian
    * [x] Chambolle Pock
    * [x] CPPA
    * [x] Douglas Rachford
    * [x] EPM
    * [x] FrankWolfe
    * [x] Levenberg Marquardt
    * [x] Nelder Mean
    * [x] particle swarm
    * [x] PD Semismoth Newton
    * [x] quasi Newton
       *  [x] direction updates 
    * [x] Stochastic Gradient Descent
    * [x] Subgradient Method
    * [x] TCG
    * [x] trust regions
  * [x] Handle State decorators
* [x] Include prints / more info (at the end?) for record/debug

### CODE REVIEW

**Comment by kellertuer** [comment]
The section about step size now reports more precisely about the step size used. Example 2 from above now prints in that section

```juliaREPL
## Stepsize
Armijolineseach() with keyword parameters
  * initial_stepsize = 1.0
  * retraction_method = ExponentialRetraction()
  * contraction_factor = 0.95
  * sufficient_decrease = 0.1
  * linesearch_stopsize = 0.0
and a last stepsize of 0.8573749999999999
```

File: src/plans/stopping_criterion.jl [hunk 1]
```diff
@@ -364,6 +464,25 @@ function (c::StopWhenAll)(p::AbstractManoptProblem, s::AbstractManoptSolverState
...
+    has_stopped = length(c.reason) > 0
+    s = has_stopped ? "reached" : "not reached"
+    r = "Stop When _all_ of the following are fulfilled:\n"
+    for cs in c.criteria
+        r = "$r    $(status_summary(cs))\n"
+    end
+    return "$(r)Overall: $s"
+end
+function indicates_convergence(c::StopWhenAll)
+    return any(indicates_convergence(ci) for ci in c.criteria)
```
Comment (mateuszbaran) [inline src/plans/stopping_criterion.jl:477]: Shouldn't it be `all`?

**Review by mateuszbaran — COMMENTED** [review]
Thanks for working on it, it will be useful for https://github.com/SciML/Optimization.jl/pull/437 .

File: src/plans/stopping_criterion.jl (hunk 1 above)
Comment (kellertuer) [reply src/plans/stopping_criterion.jl:477]: I thought so first as well. But it is any.
If this indicates to stop, then _all_ its containing ones indicate that. If only one of them indicated true, then it did converge.
For the StopWhenAny it is also not any but one has to check whether the active ones (the ones that indicated stop) have one of them who indicated stop.

So Any should be fine.

Comment (mateuszbaran) [reply src/plans/stopping_criterion.jl:477]: I see, that makes sense.

**Comment by kellertuer** [comment]
That PR started the idea, but it took a while until I found (or had the idea for) a good approach. this will also still take quite a while, since the single states for each solver might be quite some work still.

**Comment by kellertuer** [comment]
Besides a bit of code coverage in the tests, this is finished, so can be reviewed already.

File: src/plans/debug.jl [hunk 2]
```diff
@@ -69,11 +69,31 @@ function DebugSolverState(
...
+    end
+end
+function show(io::IO, dst::DebugSolverState)
+    return print(io, status_summary(dst))
+end
 dispatch_state_decorator(::DebugSolverState) = Val(true)
 
 #
-# Meta Debugs
+# Meta Debug Acrions
```
Comment (mateuszbaran) [inline src/plans/debug.jl:96]:
```suggestion
# Meta Debug Actions
```

File: src/plans/plan.jl [hunk 3]
```diff
@@ -1,3 +1,14 @@
+"""
+status_summary(e)
```
Comment (mateuszbaran) [inline src/plans/plan.jl:2]:
```suggestion
    status_summary(e)
```

File: src/plans/plan.jl [hunk 4]
```diff
@@ -1,3 +1,14 @@
...
+
+return a string reportin gabout the current status of `e`,
```
Comment (mateuszbaran) [inline src/plans/plan.jl:4]:
```suggestion
Teturn a string reporting about the current status of `e`,
```

File: src/plans/record.jl [hunk 5]
```diff
@@ -15,6 +15,7 @@ called from within [`stop_solver!`](@ref) which returns true afterwards.
 * `recorded_values` an `Array` of the recorded values.
 """
 abstract type RecordAction <: AbstractStateAction end
+status_summary(ra::RecordAction) = "$(ra)"
```
Comment (mateuszbaran) [inline src/plans/record.jl:18]: This is the same as default.

File: src/plans/record.jl [hunk 6]
```diff
@@ -275,6 +297,21 @@ getindex(r::RecordGroup, s::Symbol) = get_record(r, s)
 getindex(r::RecordGroup, s::NTuple{N,Symbol}) where {N} = get_record(r, s)
 getindex(r::RecordGroup, i) = get_record(r, i)
 
+@doc raw"""
+    RecordCost <: RecordAction
+
+record the current cost function value, see [`get_cost`](@ref).
```
Comment (mateuszbaran) [inline src/plans/record.jl:303]:
```suggestion
Record the current cost function value, see [`get_cost`](@ref).
```

File: src/plans/solver_state.jl [hunk 7]
```diff
@@ -289,3 +291,22 @@ function update_storage!(a::AbstractStateAction, d::Dict{Symbol,<:Any})
     # update keys
     return a.keys = Tuple(keys(a.values))
 end
+
+"""
+    get_count(s::AbstractManoptSolverState, :Symbol)
```
Comment (mateuszbaran) [inline src/plans/solver_state.jl:296]:
```suggestion
    get_count(ams::AbstractManoptSolverState, ::Symbol)
```

File: src/plans/solver_state.jl [hunk 8]
```diff
@@ -289,3 +291,22 @@ function update_storage!(a::AbstractStateAction, d::Dict{Symbol,<:Any})
...
+
+Obtain the count for a certain countable size, e.g. the `:Iterations`.
+This function returns 0 if there was nothing to count
+
+Available symbols from within the solver state
+
+* `:Iterations` is passed on to the `stop` field to obtain the
+  iterataion the solver stopped.
```
Comment (mateuszbaran) [inline src/plans/solver_state.jl:304]:
```suggestion
  iterataion at which the solver stopped.
```

File: src/plans/stepsize.jl [hunk 9]
```diff
@@ -194,7 +201,7 @@ mutable struct ArmijoLinesearch{TRM<:AbstractRetractionMethod,F} <: Linesearch
     linesearch_stopsize::Float64
     initial_guess::F
     function ArmijoLinesearch(
-        M;
+        M=DefaultManifold(2);
```
Comment (mateuszbaran) [inline src/plans/stepsize.jl:204]:
```suggestion
        M::AbstractManifold=DefaultManifold(2);
```

**Review by mateuszbaran — COMMENTED** [review]
This is a very nice addition but it looks like there is going to be quite a bit of merging conflicts with #212 :sweat_smile:

**Comment by kellertuer** [comment]
> This is a very nice addition but it looks like there is going to be quite a bit of merging conflicts with #212 😅

Maybe a bit. I can wait for the merge of #212 and resolve them here, no problem. I just wanted to get the whole picture in there.
Note that the show/summary functions never print the storage, I felt that might be too much information for a solver summary especially

File: src/plans/plan.jl (hunk 4 above)
Comment (kellertuer) [reply src/plans/plan.jl:4]: Did you mean to type `Return`?

File: src/plans/record.jl (hunk 5 above)
Comment (kellertuer) [reply src/plans/record.jl:18]: Is already removed in the next commit – I introduced the default a bit late. I am currently working on record code cov anyways.

File: src/plans/plan.jl (hunk 4 above)
Comment (mateuszbaran) [reply src/plans/plan.jl:4]: Yes, sorry

**Comment by mateuszbaran** [comment]
Yes, it won't be complicated merging but still in some cases same lines are touched by both PRs.

**Comment by kellertuer** [comment]
As I said I can resolve them here for sure.

**Comment by kellertuer** [comment]
You have to love GitHub Actions these days. Not only does the Windows CI randomly fail with a random `NaN` these days. Now it is also random whether codecov-reports are sent out. Not so cool if that is the main idea of current CI runs.

**Comment by kellertuer** [comment]
Since I was cuddling with code-cos anyways I checked why about 9 lines were uncovered and encountered a bug in the tests (missing copy). Now just one case in TCG (4 lines of code) remain uncovered, since that is hard to hit.

With this I declare this PR finished. We can surely wait for #212 to merge first though. Quarto #210 will wait for the same merge day so all 3 will form the next release together

**Pull request merged**